mbedtls/src/error.rs

```diff
@@ -95,59 +147,79 @@ impl From<Infallible> for Error {
     }
 }
 
+impl From<LoError> for Error {
+    fn from(error: LoError) -> Error {
+        Error::LowLevel(error)
+    }
+}
+
+impl From<HiError> for Error {
+    fn from(error: HiError) -> Error {
+        Error::HighLevel(error)
+    }
+}
+
+impl BitOr<LoError> for HiError {
+    type Output = Error;
+    fn bitor(self, rhs: LoError) -> Self::Output {
+        Error::HighAndLowLevel(self, rhs)
+    }
+}
+impl BitOr<HiError> for LoError {
+    type Output = Error;
+    fn bitor(self, rhs: HiError) -> Self::Output {
+        Error::HighAndLowLevel(rhs, self)
+    }
+}
+
+impl From<c_int> for Error {
+    fn from(x: c_int) -> Error {
+        let (high_level_code, low_level_code) = (-x & HiError::mask(), -x & LoError::mask());
+        if -x & (HiError::mask() | LoError::mask()) != -x || x >= 0 {
+            return Error::Other(x);
+        }
+        else if high_level_code == 0 {
+            return Error::LowLevel(LoError::from(low_level_code));
+        }
+        else if low_level_code == 0 {
+            return Error::HighLevel(HiError::from(high_level_code));
+        } else {
+            return Error::HighAndLowLevel(HiError::from(high_level_code), LoError::from(low_level_code))
+        }
+    }
+}
+
 impl fmt::Display for Error {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self {
             &Error::Utf8Error(Some(ref e)) => {
                 f.write_fmt(format_args!("Error converting to UTF-8: {}", e))
             }
             &Error::Utf8Error(None) => f.write_fmt(format_args!("Error converting to UTF-8")),
-            &Error::Other(i) => f.write_fmt(format_args!("mbedTLS unknown error ({})", i)),
-            &Error::__Nonexhaustive => unreachable!("__Nonexhaustive value should not be instantiated"),
-            e @ _ => f.write_fmt(format_args!("mbedTLS error {:?}", e)),
+            &Error::LowLevel(e) => f.write_fmt(format_args!("{}", e.as_str())),
+            &Error::HighLevel(e) => f.write_fmt(format_args!("{}", e.as_str())),
+            &Error::HighAndLowLevel(hi, lo) => f.write_fmt(format_args!("({}, {})", hi.as_str(), lo.as_str())),
+            &Error::Other(code) => f.write_fmt(format_args!("mbedTLS unknown error code {}", code)),
```

Comment:
Nit: shorter version (same for other branches above)
```suggestion
            &Error::Other(code) => write!("mbedTLS unknown error code {}", code),
```